forgerock-auth/src/main/java/org/forgerock/android/auth/WebAuthnDataRepository.kt

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2022 ForgeRock. All rights reserved.
+ *
+ *  This software may be modified and distributed under the terms
+ *  of the MIT license. See the LICENSE file for details.
+ */
+package org.forgerock.android.auth
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.util.Base64.*
+import org.forgerock.android.auth.Logger.Companion.error
+import org.forgerock.android.auth.Logger.Companion.warn
+import org.forgerock.android.auth.webauthn.PublicKeyCredentialSource
+import org.forgerock.android.auth.webauthn.PublicKeyCredentialSource.Companion.fromJson
+import org.json.JSONArray
+import org.json.JSONObject
+
+/**
+ * A Repository that store [PublicKeyCredentialSource]
+ */
+
+private val TAG = WebAuthnDataRepository::class.java.simpleName
+internal const val ALLOW_CREDENTIALS = "ALLOW_CREDENTIALS"
+private const val ORG_FORGEROCK_V_1_WEBAUTHN_KEYS = "org.forgerock.v1.WEBAUTHN_KEYS"
+private const val ORG_FORGEROCK_V_1_WEBAUTHN = "org.forgerock.v1.WEBAUTHN"
+private const val ORG_FORGEROCK_V_2_WEBAUTHN = "org.forgerock.v2.WEBAUTHN"
+
+open class WebAuthnDataRepository internal constructor(val context: Context,
+                                                       sharedPreferences: SharedPreferences?) {
+    private lateinit var dataRepository: SharedPreferences
+
+    init {
+        try {
+            dataRepository = sharedPreferences
+                ?: SecuredSharedPreferences(context,
+                    ORG_FORGEROCK_V_1_WEBAUTHN,
+                    ORG_FORGEROCK_V_1_WEBAUTHN_KEYS)
+            dataRepository = this.migrate(dataRepository)
+        } catch (error: Error) {
+            //Something wrong cannot support Usernameless
+            error(TAG, error, "UsernameLess cannot be supported.")
+        }
+    }
+
+    internal open fun getNewSharedPreferences(): SharedPreferences {
+        return EncryptedPreferences.getInstance(context, ORG_FORGEROCK_V_2_WEBAUTHN)
+    }
+
+    private fun migrate(sharedPreferences: SharedPreferences): SharedPreferences {
+        val encryptedPreferences = getNewSharedPreferences()
+        sharedPreferences.getString(ALLOW_CREDENTIALS, null)?.apply {
+            val map = mutableMapOf<String, MutableSet<String>>()
+            val array = JSONArray(this)
+            for (o in array.iterator<JSONObject>()) {
```

Comment:
tricky to understand this code...lets sync up tomorrow

Reply:
Put some comments in the code.

Reply:
Something like this?            

array.iterator<JSONObject>().forEach {
                fromJson(it).let { credentialSource ->
                    map[credentialSource.rpid] = map[credentialSource.rpid] ?: mutableSetOf()
                    map[credentialSource.rpid]?.add(credentialSource.toJson().put("created", System.currentTimeMillis()).toString())
                }
            }